server/src/main/java/org/elasticsearch/index/reindex/ClientScrollableHitSource.java

```diff
@@ -35,55 +35,97 @@
+import java.util.function.Function;
 
 import static java.util.Collections.emptyList;
 import static java.util.Collections.unmodifiableList;
 import static org.elasticsearch.common.unit.TimeValue.timeValueNanos;
-import static org.elasticsearch.common.util.CollectionUtils.isEmpty;
 
 /**
  * A scrollable source of hits from a {@linkplain Client} instance.
  */
 public class ClientScrollableHitSource extends ScrollableHitSource {
     private final ParentTaskAssigningClient client;
     private final SearchRequest firstSearchRequest;
+    private final String resumableSortingField;
+    private Function<SearchRequest, SearchRequest> searchRequestCloner;
 
     public ClientScrollableHitSource(Logger logger, BackoffPolicy backoffPolicy, ThreadPool threadPool, Runnable countSearchRetry,
-            Consumer<Exception> fail, ParentTaskAssigningClient client, SearchRequest firstSearchRequest) {
-        super(logger, backoffPolicy, threadPool, countSearchRetry, fail);
+                                     Consumer<AsyncResponse> onResponse, Consumer<Exception> fail,
+                                     ParentTaskAssigningClient client, SearchRequest firstSearchRequest, String resumableSortingField,
+                                     Function<SearchRequest, SearchRequest> searchRequestCloner) {
+        super(logger, backoffPolicy, threadPool, countSearchRetry, onResponse, fail, resumableSortingField);
         this.client = client;
         this.firstSearchRequest = firstSearchRequest;
+        this.resumableSortingField = resumableSortingField;
+        this.searchRequestCloner = searchRequestCloner;
     }
 
     @Override
-    public void doStart(Consumer<? super Response> onResponse) {
-        if (logger.isDebugEnabled()) {
-            logger.debug("executing initial scroll against {}",
-                    isEmpty(firstSearchRequest.indices()) ? "all indices" : firstSearchRequest.indices());
+    public void doStart(RejectAwareActionListener searchListener) {
+        client.search(firstSearchRequest, wrapListener(searchListener));
+    }
+
+    @Override
+    protected void doRestart(RejectAwareActionListener searchListener, long retryFromValue) {
+        SearchRequest retryFromRequest = createRetryFromRequest(retryFromValue);
+        client.search(retryFromRequest, wrapListener(searchListener));
+    }
+
+    private SearchRequest createRetryFromRequest(long retryFromValue) {
+        SearchRequest retryFromRequest = searchRequestCloner.apply(firstSearchRequest);
+        RangeQueryBuilder rangeQueryBuilder = new RangeQueryBuilder(resumableSortingField).gte(retryFromValue);
+        if (retryFromRequest.source() == null) {
+            retryFromRequest.source(new SearchSourceBuilder().query(rangeQueryBuilder));
+        } else if (retryFromRequest.source().query() == null) {
+            retryFromRequest.source().query(rangeQueryBuilder);
+        } else {
+            throw new UnsupportedOperationException("not yet implemented");
```

Comment:
perhaps wrap original query using BoolQueryBuilder where the rangeQueryBuilder is added as filter and the original query as must.